File: src/main/java/dev.toad/Async.java
package dev.toad;

import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Supplier;

public class Async {

  public static ScheduledExecutorService executor =
    Executors.newScheduledThreadPool(32);

  public static class LoopHandle implements AutoCloseable {

    CompletableFuture<Object> fut = null;
    final AtomicBoolean break_ = new AtomicBoolean(false);
    final Runnable onStop;

    LoopHandle(Runnable onStop) {
      this.onStop = onStop;
    }

    public void join() throws Throwable, InterruptedException {
      try {
        this.fut.get();
      } catch (ExecutionException e) {
        throw e.getCause();
      }
    }

    public void stop() throws Throwable, InterruptedException {
      this.break_.set(true);
      this.join();
      this.onStop.run();
    }

    @Override
    public void close() {
      try {
        this.stop();
      } catch (Throwable t) {}
    }
  }

  public static LoopHandle loop(Runnable f) {
    return Async.loop(f, () -> {});
  }

  public static LoopHandle loop(Runnable f, Runnable onStop) {
    var handle = new LoopHandle(onStop);
    var fut = Async.pollCompletable(() -> {
      if (handle.break_.get()) {
        return Optional.of(new Object());
      } else {
        f.run();
        return Optional.empty();
      }
    });

    handle.fut = fut;

    return handle;
  }

  public static <T> CompletableFuture<T> pollCompletable(
    Supplier<Optional<T>> sup
  ) {
    var fut = new CompletableFuture();
    var pollTask = Async.executor.scheduleAtFixedRate(
      () -> {
        try {
          var t = sup.get();
          if (!t.isEmpty()) {
            fut.complete(t.get());
          }
        } catch (Throwable ex) {
          fut.completeExceptionally(ex);
        }
      },
      0,
      10,
      TimeUnit.MILLISECONDS
    );

    return fut.whenComplete((t, err) -> {
      pollTask.cancel(true);
    });
  }
}

File: src/test/scala/Async.scala
import dev.toad.Async
import java.util.Date
import java.util.Optional
import java.util.concurrent.TimeUnit
import java.util.concurrent.CompletableFuture
import java.time.Duration

class AsyncTest extends munit.FunSuite {
  test("loop() runs until LoopHandle.stop() invoked") {
    var counter = 0;
    val loop = Async.loop(() => {
      counter += 1
    });

    val start = Date().getTime();
    while (counter < 10 && Date().getTime() < start + 1000) {
      Thread.sleep(Duration.ofMillis(10))
    }

    loop.stop()
    assert(counter >= 10)
  }

  test("LoopHandle.stop() prevents further runs") {
    var stopped = false;
    val loop = Async.loop(() => {
      if stopped then throw new Error()
    });

    loop.stop()
    stopped = true
  }

  test("LoopHandle.join() rethrows exceptions") {
    case class LoopThrew() extends Throwable {}
    case class LoopDidNotThrow() extends Throwable {}

    try {
      val loop = Async.loop((() => {
        throw LoopThrew()
      }): java.lang.Runnable)

      loop.join();
      throw LoopDidNotThrow()
    } catch {
      case LoopThrew() => {}
      case e           => throw e
    }
  }

  test("pollCompletable polls then completes") {
    var polls = 0

    val fut = Async.pollCompletable(() => {
      polls = polls + 1
      if polls == 10 then {
        Optional.of(true)
      } else {
        Optional.empty
      }
    })

    fut.get(200, TimeUnit.MILLISECONDS)
  }

  test("pollCompletable does not poll after completion") {
    var polls = 0
    val fut = Async.pollCompletable(() => {
      polls = polls + 1
      Optional.of(true)
    })

    fut.get()

    Thread.sleep(100)
    assertEquals(polls, 1)
  }

  test("pollCompletable completesExceptionally when poll throws") {
    val err = Async
      .pollCompletable[Object](() => throw Exception("foo"))
      .handle((ok, e) => {
        if ok != null then {
          "CompletableFuture was completed without exception"
        } else {
          e.getMessage
        }
      })
      .get()

    assertEquals(err, "java.lang.Exception: foo")
  }
}
